File: packages/core/src/templates.ts
import type { ExpressionString } from "./types";

export interface RuleTemplate {
  params: string[];
  expression: ExpressionString;
}

export const RULE_TEMPLATES: Record<string, RuleTemplate> = {
  minValue: {
    params: ["fieldId", "minValue"],
    expression: "value('${fieldId}') < ${minValue} && value('${fieldId}') !== null && value('${fieldId}') !== ''",
  },
  maxValue: {
    params: ["fieldId", "maxValue"],
    expression: "value('${fieldId}') > ${maxValue}",
  },
  between: {
    params: ["fieldId", "min", "max"],
    expression: "(value('${fieldId}') < ${min} || value('${fieldId}') > ${max}) && value('${fieldId}') !== null",
  },
  fieldEquals: {
    params: ["fieldId", "value"],
    expression: "value('${fieldId}') === '${value}'",
  },
  fieldNotEquals: {
    params: ["fieldId", "value"],
    expression: "value('${fieldId}') !== '${value}'",
  },
  isEmpty: {
    params: ["fieldId"],
    expression: "isEmpty(value('${fieldId}'))",
  },
  isNotEmpty: {
    params: ["fieldId"],
    expression: "!isEmpty(value('${fieldId}'))",
  },
  fieldsEqual: {
    params: ["fieldId1", "fieldId2"],
    expression: "value('${fieldId1}') === value('${fieldId2}')",
  },
  fieldGreaterThan: {
    params: ["fieldId1", "fieldId2"],
    expression: "value('${fieldId1}') > value('${fieldId2}')",
  },
  fieldLessThan: {
    params: ["fieldId1", "fieldId2"],
    expression: "value('${fieldId1}') < value('${fieldId2}')",
  },
  matches: {
    params: ["fieldId", "pattern"],
    expression: "/${pattern}/.test(value('${fieldId}') || '')",
  },
  minLength: {
    params: ["fieldId", "minLength"],
    expression: "(value('${fieldId}') || '').length < ${minLength}",
  },
  maxLength: {
    params: ["fieldId", "maxLength"],
    expression: "(value('${fieldId}') || '').length > ${maxLength}",
  },
  includes: {
    params: ["fieldId", "value"],
    expression: "(value('${fieldId}') || []).includes('${value}')",
  },
  notIncludes: {
    params: ["fieldId", "value"],
    expression: "!(value('${fieldId}') || []).includes('${value}')",
  },
};

/**
 * Register a custom rule template. Templates are named, parameterized
 * expression generators — a host can add domain-specific shorthands
 * (typically delegating to a function added via registerFunction).
 * Register the same template in every engine that prepares the schema.
 */
export function registerTemplate(name: string, template: RuleTemplate): void {
  if (!/^[a-zA-Z_]\w*$/.test(name)) {
    throw new Error(`Invalid template name: ${name}`);
  }
  RULE_TEMPLATES[name] = template;
}

export function unregisterTemplate(name: string): void {
  delete RULE_TEMPLATES[name];
}

export function resolveTemplate(
  template: RuleTemplate,
  params: Record<string, any>,
  templateId?: string
): ExpressionString {
  let expression = template.expression;

  for (const paramName of template.params) {
    if (!(paramName in params)) {
      throw new Error(
        `Missing required parameter "${paramName}"${templateId ? ` for template "${templateId}"` : ""}`
      );
    }
  }

  for (const [paramName, paramValue] of Object.entries(params)) {
    const paramPattern = new RegExp(
      `\\$\\{${paramName.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\}`,
      "g"
    );

    let valueStr: string;
    if (typeof paramValue === "string") {
      valueStr = paramValue;
    } else if (typeof paramValue === "number" || typeof paramValue === "boolean") {
      valueStr = String(paramValue);
    } else {
      valueStr = JSON.stringify(paramValue);
    }

    // Function replacement — a string replacement would interpret `$&`-style
    // sequences inside param values (e.g. regex patterns) as substitutions.
    expression = expression.replace(paramPattern, () => valueStr);
  }

  return expression;
}

export function isTemplateReference(exprString: string): boolean {
  if (RULE_TEMPLATES[exprString]) return true;

  const templateCallMatch = exprString.match(/^(\w+)\(/);
  if (!templateCallMatch || RULE_TEMPLATES[templateCallMatch[1]] === undefined) {
    return false;
  }

  // A template reference is exactly `name(args)` — the paren opened after the
  // name must close at the END of the string. Compound expressions that merely
  // START with a template name (e.g. "isEmpty(x) && value('y') === 'z'") are
  // plain expressions, not template calls.
  let depth = 0;
  let quote: string | null = null;
  for (let i = templateCallMatch[1].length; i < exprString.length; i++) {
    const ch = exprString[i];
    if (quote) {
      if (ch === quote) quote = null;
    } else if (ch === "'" || ch === '"') {
      quote = ch;
    } else if (ch === "(") {
      depth++;
    } else if (ch === ")") {
      depth--;
      if (depth === 0) return i === exprString.length - 1;
    }
  }
  return false;
}

/** Split template args on commas, ignoring commas inside quoted strings
 * (e.g. matches('^[A-Z]{2,4}$') is a single argument). */
function splitTemplateArgs(paramsString: string): string[] {
  const args: string[] = [];
  let current = "";
  let quote: string | null = null;
  for (const ch of paramsString) {
    if (quote) {
      current += ch;
      if (ch === quote) quote = null;
    } else if (ch === "'" || ch === '"') {
      quote = ch;
      current += ch;
    } else if (ch === ",") {
      args.push(current.trim());
      current = "";
    } else {
      current += ch;
    }
  }
  if (current.trim()) args.push(current.trim());
  return args;
}

export function parseTemplateReference(exprString: string): { templateId: string; params: Record<string, any> } {
  if (RULE_TEMPLATES[exprString]) {
    return { templateId: exprString, params: {} };
  }

  const match = exprString.match(/^(\w+)\((.*)\)$/);
  if (!match) {
    throw new Error(`Invalid template reference syntax: "${exprString}"`);
  }

  const templateId = match[1];
  const paramsString = match[2].trim();

  const template = RULE_TEMPLATES[templateId];
  if (!template) {
    throw new Error(`Unknown template: "${templateId}". Available: ${Object.keys(RULE_TEMPLATES).join(", ")}`);
  }

  const params: Record<string, any> = {};

  if (paramsString) {
    const paramValues = splitTemplateArgs(paramsString);

    const shouldAutoInferFieldId =
      template.params[0] === "fieldId" &&
      paramValues.length < template.params.length;

    let valueIndex = 0;
    for (let i = 0; i < template.params.length; i++) {
      const paramName = template.params[i];

      if (shouldAutoInferFieldId && paramName === "fieldId") {
        continue;
      }

      if (valueIndex < paramValues.length) {
        let value = paramValues[valueIndex].trim();

        if (/^\d+(\.\d+)?$/.test(value)) {
          params[paramName] = parseFloat(value);
        } else if (/^['"].*['"]$/.test(value)) {
          params[paramName] = value.slice(1, -1);
        } else {
          params[paramName] = value;
        }

        valueIndex++;
      }
    }
  }

  return { templateId, params };
}

function resolveExpressionString(exprString: string, rule: any): string {
  if (isTemplateReference(exprString)) {
    const { templateId, params } = parseTemplateReference(exprString);
    const template = RULE_TEMPLATES[templateId];

    if (template.params.includes("fieldId") && !params.fieldId) {
      if (rule.affects && rule.affects.length > 0 && rule.affects[0].target) {
        params.fieldId = rule.affects[0].target;
      } else {
        throw new Error(
          `Rule "${rule.id}" uses template "${templateId}" which requires a fieldId, but no fieldId param was provided and no target found in affects`
        );
      }
    }

    return resolveTemplate(template, params, templateId);
  }

  return exprString;
}

function resolveExpression(expr: any, rule: any): any {
  if (typeof expr === "string") {
    return resolveExpressionString(expr, rule);
  }

  if (Array.isArray(expr)) {
    return expr.map(e => resolveExpression(e, rule));
  }

  if (typeof expr === "object" && expr !== null) {
    if ("expressions" in expr) {
      return {
        ...expr,
        expressions: resolveExpression(expr.expressions, rule),
      };
    }
  }

  return expr;
}

export function resolveRuleTemplates(rules: any[]): any[] {
  return rules.map((rule) => {
    let when = rule.when;

    if (typeof when === "string") {
      when = { expressions: when };
    }

    return {
      ...rule,
      when: resolveExpression(when, rule),
    };
  });
}

File: packages/python/protoform/templates.py
"""
Built-in rule templates — Python port of @protoform/core templates.
"""

from __future__ import annotations

import re
from typing import Any


RULE_TEMPLATES: dict[str, dict] = {
    "minValue": {
        "params": ["fieldId", "minValue"],
        "expression": "value('${fieldId}') < ${minValue} && value('${fieldId}') !== null && value('${fieldId}') !== ''",
    },
    "maxValue": {
        "params": ["fieldId", "maxValue"],
        "expression": "value('${fieldId}') > ${maxValue}",
    },
    "between": {
        "params": ["fieldId", "min", "max"],
        "expression": "(value('${fieldId}') < ${min} || value('${fieldId}') > ${max}) && value('${fieldId}') !== null",
    },
    "fieldEquals": {
        "params": ["fieldId", "value"],
        "expression": "value('${fieldId}') === '${value}'",
    },
    "fieldNotEquals": {
        "params": ["fieldId", "value"],
        "expression": "value('${fieldId}') !== '${value}'",
    },
    "isEmpty": {
        "params": ["fieldId"],
        "expression": "isEmpty(value('${fieldId}'))",
    },
    "isNotEmpty": {
        "params": ["fieldId"],
        "expression": "!isEmpty(value('${fieldId}'))",
    },
    "fieldsEqual": {
        "params": ["fieldId1", "fieldId2"],
        "expression": "value('${fieldId1}') === value('${fieldId2}')",
    },
    "fieldGreaterThan": {
        "params": ["fieldId1", "fieldId2"],
        "expression": "value('${fieldId1}') > value('${fieldId2}')",
    },
    "fieldLessThan": {
        "params": ["fieldId1", "fieldId2"],
        "expression": "value('${fieldId1}') < value('${fieldId2}')",
    },
    "matches": {
        "params": ["fieldId", "pattern"],
        "expression": "/${pattern}/.test(value('${fieldId}') || '')",
    },
    "minLength": {
        "params": ["fieldId", "minLength"],
        "expression": "(value('${fieldId}') || '').length < ${minLength}",
    },
    "maxLength": {
        "params": ["fieldId", "maxLength"],
        "expression": "(value('${fieldId}') || '').length > ${maxLength}",
    },
    "includes": {
        "params": ["fieldId", "value"],
        "expression": "(value('${fieldId}') || []).includes('${value}')",
    },
    "notIncludes": {
        "params": ["fieldId", "value"],
        "expression": "!(value('${fieldId}') || []).includes('${value}')",
    },
}

import re as _re_name


def register_template(name: str, params: list[str], expression: str) -> None:
    """Register a custom rule template — a named, parameterized expression
    generator. Hosts add domain-specific shorthands (typically delegating to
    a function added via register_function). Register the same template in
    every engine that prepares the schema."""
    if not _re_name.match(r"^[a-zA-Z_]\w*$", name):
        raise ValueError(f"Invalid template name: {name}")
    RULE_TEMPLATES[name] = {"params": list(params), "expression": expression}


def unregister_template(name: str) -> None:
    RULE_TEMPLATES.pop(name, None)


def resolve_template(template_id: str, params: dict[str, Any]) -> str:
    """Resolve a template by substituting parameter values into the expression."""
    tpl = RULE_TEMPLATES.get(template_id)
    if tpl is None:
        raise ValueError(f"Unknown template: {template_id}")

    expression = tpl["expression"]
    for param_name in tpl["params"]:
        if param_name not in params:
            raise ValueError(f"Missing parameter '{param_name}' for template '{template_id}'")

    for key, val in params.items():
        # Literal replacement — regex-based substitution corrupts params
        # containing backslashes (e.g. matches('^\d{4}$')).
        expression = expression.replace("${" + key + "}", str(val))

    return expression


def is_template_reference(expr_string: str) -> bool:
    """Check if an expression string is a template reference.

    A template reference is exactly ``name`` or ``name(args)`` — the paren
    opened after the name must close at the END of the string. Compound
    expressions that merely start with a template name (e.g.
    ``isEmpty(x) && value('y') === 'z'``) are plain expressions.
    """
    if expr_string in RULE_TEMPLATES:
        return True
    match = re.match(r"^(\w+)\(", expr_string)
    if not match or match.group(1) not in RULE_TEMPLATES:
        return False

    depth = 0
    quote: str | None = None
    for i in range(len(match.group(1)), len(expr_string)):
        ch = expr_string[i]
        if quote:
            if ch == quote:
                quote = None
        elif ch in ("'", '"'):
            quote = ch
        elif ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
            if depth == 0:
                return i == len(expr_string) - 1
    return False


def _split_template_args(params_str: str) -> list[str]:
    """Split template args on commas, ignoring commas inside quoted strings
    (e.g. matches('^[A-Z]{2,4}$') is a single argument)."""
    args: list[str] = []
    current = ""
    quote: str | None = None
    for ch in params_str:
        if quote:
            current += ch
            if ch == quote:
                quote = None
        elif ch in ("'", '"'):
            quote = ch
            current += ch
        elif ch == ",":
            args.append(current.strip())
            current = ""
        else:
            current += ch
    if current.strip():
        args.append(current.strip())
    return args


def parse_template_reference(expr_string: str) -> tuple[str, dict[str, Any]]:
    """Parse 'templateName(arg1, arg2)' into (template_id, params)."""
    if expr_string in RULE_TEMPLATES:
        return expr_string, {}

    match = re.match(r"^(\w+)\((.*)\)$", expr_string)
    if not match:
        raise ValueError(f"Invalid template reference: {expr_string}")

    template_id = match.group(1)
    params_str = match.group(2).strip()

    tpl = RULE_TEMPLATES.get(template_id)
    if tpl is None:
        raise ValueError(f"Unknown template: {template_id}")

    params: dict[str, Any] = {}
    if params_str:
        param_values = _split_template_args(params_str)

        # Auto-infer fieldId
        should_auto_infer = (
            tpl["params"][0] == "fieldId"
            and len(param_values) < len(tpl["params"])
        )

        value_index = 0
        for i, param_name in enumerate(tpl["params"]):
            if should_auto_infer and param_name == "fieldId":
                continue
            if value_index < len(param_values):
                raw = param_values[value_index]
                # Try numeric
                if re.match(r"^-?\d+(\.\d+)?$", raw):
                    params[param_name] = float(raw) if "." in raw else int(raw)
                elif (raw.startswith("'") and raw.endswith("'")) or (raw.startswith('"') and raw.endswith('"')):
                    params[param_name] = raw[1:-1]
                else:
                    params[param_name] = raw
                value_index += 1

    return template_id, params


def resolve_expression_string(expr_string: str, rule: dict) -> str:
    """Resolve a single expression string (may be a template reference)."""
    if not is_template_reference(expr_string):
        return expr_string

    template_id, params = parse_template_reference(expr_string)
    tpl = RULE_TEMPLATES[template_id]

    if "fieldId" in tpl["params"] and "fieldId" not in params:
        affects = rule.get("affects", [])
        if affects:
            params["fieldId"] = affects[0].get("target", "")
        else:
            raise ValueError(
                f"Rule '{rule.get('id')}' uses template '{template_id}' which requires fieldId"
            )

    return resolve_template(template_id, params)


def _resolve_expr(expr: Any, rule: dict) -> Any:
    """Recursively resolve template references in an expression tree."""
    if isinstance(expr, str):
        return resolve_expression_string(expr, rule)

    if isinstance(expr, list):
        return [_resolve_expr(e, rule) for e in expr]

    if isinstance(expr, dict) and "expressions" in expr:
        return {**expr, "expressions": _resolve_expr(expr["expressions"], rule)}

    return expr


def resolve_rule_templates(rules: list[dict]) -> list[dict]:
    """Resolve all template references in a list of rules."""
    resolved = []
    for rule in rules:
        when = rule.get("when", "")
        if isinstance(when, str):
            when = {"expressions": when}
        resolved.append({**rule, "when": _resolve_expr(when, rule)})
    return resolved
